### PR TITLE
Fix (Core) issue in Creature::UpdateEntry

### PR DESCRIPTION
**Pull request by acidmanifesto**
This was a devil to debug. This should now load the scriptid if any and then force clear any stuck menu status before preparing and sending a gossip to a player.

## Changes Proposed:
-  We must update last scriptId or it looks like we reloaded a script, breaking some things such as gossip temporarily.

## Issues Addressed:

- Closes https://github.com/chromiecraft/chromiecraft/issues/2604
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/9745

## Tests Performed:

- Builds and Tested in Game.

## How to Test the Changes:

1. .go c 37605
2. run and right-click on vendor

image

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Review by Nefertumm — APPROVED** [review]

**Comment by IntelligentQuantum** [comment]
Screenshot_20211227_093105

@acidmanifesto still bug

**Comment by IntelligentQuantum** [comment]
Screenshot_20211227_093105

@acidmanifesto still bug

File: src/server/game/Entities/Creature/Creature.cpp [hunk 1]
```diff
@@ -549,6 +549,10 @@ bool Creature::UpdateEntry(uint32 Entry, const CreatureData* data, bool changele
     SetDetectionDistance(cInfo->detection_range);
 
     LoadSpellTemplateImmunity();
+
+    // We must update last scriptId or it looks like we reloaded a script, breaking some things such as gossip temporarily
+    LastUsedScriptID = GetScriptId();
+
```
Comment (Kitzunu) [inline src/server/game/Entities/Creature/Creature.cpp:555]: To me this just looks weird. We don't pass this variable anywhere after it. So why would it make a difference?

Comment (acidmanifesto) [reply src/server/game/Entities/Creature/Creature.cpp:555]: https://github.com/azerothcore/azerothcore-wotlk/blob/8497a3683023166cab1123751575ca3131ce08a2/src/server/game/Entities/Creature/Creature.cpp#L1045

It is the same meathod that is used on creature create that enables it to load its script.

**Comment by acidmanifesto** [comment]
> Screenshot_20211227_093105
> 
> @acidmanifesto still bug

What adjustments do you have out of coursity, when i click the npc it goes straight to the vendor menu and not a gossip select? Could help me narrow down a few things.

**Comment by acidmanifesto** [comment]
@IntelligentQuantum if u can try again. if this doesnt work then i will have to see about something in the opcodes.cpp.

**Comment by IntelligentQuantum** [comment]
@acidmanifesto still bug.

>  when i click the npc it goes straight to the vendor menu and not a gossip select? 

idk really 

Os: Arch
hash https://github.com/azerothcore/azerothcore-wotlk/commit/dee19a030fd066ba1aa9868541873c5b06c2dd43
clean db
wow: enUS

**Comment by tkn963** [comment]
I can't seem to open the vendor window either. Standing still and interacting works, but opening the gossip window when moving does not let me click the 'I want to browse your goods' option.

**Comment by claudiodfc** [comment]
Isn't this fixed by #9706 ? Does this NPC have force gossip flags?

**Comment by acidmanifesto** [comment]
I will need to rebuild my whole db right jow just to see where the inconsistency is happening with me. As i dont get a gossip menu at all. I go straight to the vendor menu.

**Comment by Kitzunu** [comment]
> Isn't this fixed by #9706 ? Does this NPC have force gossip flags?

that PR is a hack and a half. Bad.

**Comment by acidmanifesto** [comment]
Ok. i have rebuild the database and everything from scratch and fresh on master. This is what i have going on.:
 I do **NOT** get a gossip menu with a vendor menu.
  I **DO** get  a vendor menu when i click on the npc while moving.
  I **did** have a issue with the npc not showing me anything while i right click on them while moving **BEFORE this PR**.
  I do **NOT** have the issue anymore after this **PR.**

At this point i will have to mark this as help wanted. Even though the issue is fixed for me, it is not fix for others.
My Enviroment is as follows:

Windows 10 Pro x64
MySQL 5.7
Boost 1.71

**Comment by tkn963** [comment]
No matter if I use a fresh database or not, the result is always the same for me. If it matters at all, I'm running Debian 11 with MariaDB 10.5.12 (MySQL 8).

**Comment by dantav** [comment]
I would like to help but I cannot find a way to constantly get the issue to happen. I tried the above and I always get a gossip menu working. Sometimes randomly gossip menu appears and I cannot click on anything. At one point on my production server I constantly got the issue to occur when moving and trying to gossip, it felt like a range miscalculation issue. However, now i'm unable to repeat this on dev or prod. It's extremely inconsistent. This is a tough bug to crack.

**Comment by acidmanifesto** [comment]
Closing due to possible solution with PR https://github.com/azerothcore/azerothcore-wotlk/pull/9962

**Pull request closed without merging**